docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,332 @@
+
+#### Minimum Request Compression Size Bytes 
+* Type: Int
+* Default: 10,240 (10KB)
+* Range: 0 - 10,485,760 (0MB - 10MB) 
+* Notes: Setting will be ignored if payload is a stream
+
+| Description                  | Client configuration             | JVM system property                  | Environment variable                     | AWS profile key                      |
+|------------------------------|----------------------------------|--------------------------------------|------------------------------------------|--------------------------------------|
+| Min payload size to compress | `requestMinCompressionSizeBytes` | `aws.requestMinCompressionSizeBytes` | `AWS_REQUEST_MIN_COMPRESSION_SIZE_BYTES` | `request_min_compression_size_bytes` |
+
+This feature can also be configured to use user supplied compression algorithms. The user can add compression algorithms using the client configuration function `addCompressionAlgorithms`. A user supplied algorithm may override the built-in implementation, and the compression algorithm must implement the `CompressionAlgorithm` interface.
+```kotlin
+val client = S3Client.fromEnvironment {
+    region = "us-west-2"
+}
+client.addCompressionAlgorithm(...)
+```
+
+### The Trait
+Below is an abbreviated example of the requestCompression [trait](https://smithy.io/2.0/spec/behavior-traits.html?highlight=requestcompression#requestcompression-trait) in use:
+
+```smithy
+@requestCompression(
+    encodings: ["gzip", "custom"]
+)
+operation ...
+```
+
+The SDK has its own list of supported compression algorithms as a list of `CompressionAlgorithm`s in the client config.
+
+```kotlin
+private val compressionAlgorithms: List<CompressionAlgorithm> = listOf(
+    GzipCompressionAlgorithm()
+)
+```
+
+### Behavior
+Each algorithm will be attempted in the order they appear in the trait until one that is supported by the runtime is found. If no algorithm is found the request will continue without compression applied.
+
+After compression occurs, the `Content-Encoding` header will contain the name of the compression algorithm used. It will be appended to the list of content encodings if the header was already present with some value.
+
+```http request
+// There was no encoding in header before compression
+POST /upload-data HTTP/1.1
+Host: example.com
+Content-Type: application/json
+Content-Encoding: gzip
+
+// There was an encoding in header before compression
+POST /upload-data HTTP/1.1
+Host: example.com
+Content-Type: application/json
+Content-Encoding: br, gzip 
+```
+
+### Compression Algorithms
+Given that Kotlin is multiplatform there should be multiple implementations of each compression algorithm. As of the time of writing this doc (11/07/2023) gzip is the only supported compression algorithm by the server side and the only one supported for
```

Comment:
simplification: "As of the time of writing this doc (11/07/2023)..." --> "Currently gzip is the only supported compression algorithm."